google-cloud-clients/google-cloud-firestore/src/test/java/com/google/cloud/firestore/it/ITSystemTest.java

```diff
@@ -889,95 +893,184 @@ public void onEvent(
-                        }
-                      } catch (Exception e) {
-                        fail(e.getMessage());
-                      }
-                      semaphore.release();
-                    }
-                  });
-
-      final boolean tryAcquire = semaphore.tryAcquire(6, 60, TimeUnit.SECONDS);
-
-      final Joiner j = Joiner.on(", ");
-      final String expectedString = j.join(expectedEvents);
-      final String actualString = j.join(actualEvents);
-      assertTrue(
-          String.format(
-              "did not receive all expected events within the deadline.%n"
-                  + "expectedEvents = [%s]%n"
-                  + "  actualEvents = [%s]%n",
-              expectedString, actualString),
-          tryAcquire);
+          query.addSnapshotListener(
+              new EventListener<QuerySnapshot>() {
+                @Override
+                public void onEvent(
+                    @Nullable QuerySnapshot value, @Nullable FirestoreException error) {
+                  receivedEvents.add(new ListenerEvent(value, error));
+                  eventsCDL.countDown();
+                }
+              });
+
+      // Perform a series of operations on some documents in a separate thread
+      // While we listen for the changes
+      updatesExecutor.submit(
+          new Runnable() {
+            DocumentReference doc1 = randomColl.document("doc1");
+
+            @Override
+            public void run() {
+              try {
+                // create the first document
+                doc1.set(map("baz", "foo")).get();
+                // update a field in the document
+                doc1.set(map("foo", "bar")).get();
+                // add a field to the document
+                doc1.set(map("foo", "bar", "bar", "foo")).get();
+                // delete the document
+                doc1.delete().get();
+              } catch (InterruptedException | ExecutionException e) {
+                fail(String.format("Error while processing doc1: %s", e.getMessage()));
+              } finally {
+                doc1CDL.countDown();
+              }
+            }
+          });
+
+      updatesExecutor.submit(
+          new Runnable() {
+            DocumentReference doc2 = randomColl.document("doc2");
+
+            @Override
+            public void run() {
+              try {
+                // create a second document
+                doc2.set(map("foo", "bar")).get();
+                // update a field in the document
+                doc2.set(map("foo", "foo")).get();
+              } catch (InterruptedException | ExecutionException e) {
+                fail(String.format("Error while processing doc2: %s", e.getMessage()));
+              } finally {
+                doc2CDL.countDown();
+              }
+            }
+          });
+
+      // Wait for the document update operations to be performed
+      final boolean doc1CDLAwait = doc1CDL.await(10, TimeUnit.SECONDS);
```

Comment:
If I am not mistaken, doc1CDLAwait and doc2CDLAwait could be a single count down latch initialized with `new CountDownLatch(2)`.

Reply:
It is true we could do this, but then we would lose the info about which of the document operations timed out for the assertion. I'd like to keep the extra detail in case the test ever fails due to this.